### ISSUE TITLE
After decoding through vaapi+ffmpeg, we get "AVFrame frame->data[3];" After., how to render with 0 copy via SDL

### ISSUE THREAD

**Issue by Mangomm** [libsdl-org/SDL#8302]
base https://github.com/libsdl-org/SDL/pull/4909.

Hello, is there a better way to deal with this problem in SDL3? I have also encountered this problem at present. The code you gave, I have a question, how is your decoded frame data uploaded to the texture?
For example, after I decode the frame using ffmpeg+vaapi, the decoded frame data is saved in "AVFrame frame->data[3];". How is the data uploaded to pTexture?

Recently, I have been studying using other solutions to render, such as qt or opengles, but due to project reasons, I think it is better to use SDL, but I am not sure how to deal with SDL at present

**Comment by slouken** [comment]
I've added an example for SDL3 of how to do cross-platform hardware decoding with ffmpeg.

Enjoy!

**Comment by Mangomm** [comment]
I looked at it and should be the address at https://github.com/slouken/SDL/blob/873bcbfecf7243523e64f544f6062d94c?

Love you, it's really exciting and that's why I've always loved developing with sdl

**Comment by slouken** [comment]
> I looked at it and should be the address at https://github.com/slouken/SDL/blob/873bcbfecf7243523e64f544f6062d94c?

No, it's already been merged to the main SDL repo:
https://github.com/libsdl-org/SDL

**Comment by Mangomm** [comment]
Hi, I'm testing the testffmpeg.c example you updated yesterday.
Setting:
Since CMakeLists.txt in the SDL/test directory did not compile successfully, I set up my own test environment using qt.
In testffmpeg.c, I use the #define HAVE_EGL configuration.
Terminal printing:
INFO: Created renderer opengles2
INFO: Video stream: h264 1280x720
INFO: Using vaapi hardware acceleration with pixel format vaapi
INFO: Audio stream: aac 2 channels, 48000 Hz

There are some problems:
1. When I wanted to test 4K(3840x2160) video, I modified:
#define WINDOW_WIDTH  3840
#define WINDOW_HEIGHT 2160
But it doesn't work, and it always displays at the resolution of the original video.

2. Compared with vlc(VA-API decoder device for X11), the SDL3 screen is fuzzy, is it missing some parameter Settings of egl?
Focus on the time section in red in the image below, which is clearly blurry.

![image](https://github.com/libsdl-org/SDL/assets/58781546/8ec15655-7c8b-4bde-9672-5cab6b267704)


In fact, I am more concerned about the first question!

**Comment by slouken** [comment]
Oh, that's just because of this:
https://github.com/libsdl-org/SDL/blob/4368f70ff9447f5960e1afa7a2bb878414c50bc1/test/testffmpeg.c#L384
You can remove that line if you want to have the video stretched to the window size.

**Comment by slouken** [comment]
I think what you're seeing is that VLC is doing smoothing. 

Does this give the same output?
https://github.com/libsdl-org/SDL/commit/bf64fecf19f87d54af2f2e3b4430318552ac997d

**Comment by Mangomm** [comment]
> 我想你看到的是 VLC 正在做平滑处理。
> 
> 这会给出相同的输出吗？ [BF64FEC](https://github.com/libsdl-org/SDL/commit/bf64fecf19f87d54af2f2e3b4430318552ac997d)


Yes, add "SDL_SetHint(SDL_HINT_RENDER_SCALE_QUALITY, "linear");"  After it successfully removed the blur.
However, the picture quality is still a little different from vlc, such as the brightness of vlc is better and looks more comfortable.

**Comment by slouken** [comment]
Do you have a link to that video for me to take a look?

**Comment by Mangomm** [comment]
I added my test video at the link below, you can download the video.
Then you can use testffmpeg.c to play the video, and then use vlc to play the video and compare the quality.
You will find that the picture quality of vlc is better, more obvious in 4k resolution, this problem is actually a few years ago I encountered, I think this is sdl do not do enough, but I do not know how to optimize, may be some parameters of the reason.

I will delete this link after you download the video.
link：
https://www.yuque.com/tashenzhibuyuanjiaowoyishengfanpeixi/nnm2gm/qdp92q9q1tdywv0c

**Comment by slouken** [comment]
I'm not able to download that video.

**Comment by Mangomm** [comment]
I used ffmpeg to crop a portion of this video (10s in length) for easy uploading to github. You can test this out, github should be able to download videos?


https://github.com/libsdl-org/SDL/assets/58781546/11b34393-7666-41fb-b785-bee53203e9e4

**Comment by Mangomm** [comment]
Hello, a new question. When I wrap testffmpeg.c as a dynamic library so (like wrapping a player), a qt application calls the dynamic library.
Create a window using an existing handle (i.e. a qt window handle or a C# window handle). If opengl is used, it will report an error.
testffmpeg.c just changes the code that creates the window:
    if (SDL_CreateWindowAndRenderer(1920, 1080, SDL_WINDOW_HIDDEN | SDL_WINDOW_RESIZABLE | SDL_WINDOW_HIGH_PIXEL_DENSITY | SDL_WINDOW_OPENGL, &_window, &_renderer) < 0) {
        return false;
    }

Replaced with:

    SDL_SetHint("SDL_HINT_VIDEO_WINDOW_SHARE_PIXEL_FORMAT", "1");
    _window = SDL_CreateWindowFrom(_hwnd);
    if(!_window){
        return false;
    }

    _renderer = SDL_CreateRenderer(_window, NULL, SDL_RENDERER_ACCELERATED);
    if(!_renderer){
        SDL_Log("Created renderer failed, %s\n", SDL_GetError());
        return false;
    }

An error occurred while SDL_CreateRenderer, printed as follows:
INFO: Created renderer failed, Couldn't find matching render driver

How do I render opengl using SDL_CreateWindowFrom to create Windows?

note: 
_hwnd is a handle of type void *, passed in from qt:
_opengl_wgt = new QOpenGLWidget();
_hwnd = (void*)_opengl_wgt->winId();

**Comment by slouken** [comment]
Please create a new issue for this, and include a standalone repro case that we can look at.

Thanks!

**Comment by Mangomm** [comment]
> Please create a new issue for this, and include a standalone repro case that we can look at.
> 
> Thanks!

ok, I'll make an example in two days and then submit a new issue

**Comment by Mangomm** [comment]
Hello, can  testffmpeg.c example support the processing of sdl2? Because in the actual development process, I found that it would be very troublesome to update the interface of sdl3, especially the audio process. I used the callback method of sdl similar to ffplay to do a lot of audio and video synchronization processing. This has forced me to spend more time thinking about these solutions, and I am sure there will be others who will encounter such problems in the future
